src/python/WMCore/Services/DDM/DDM.py

```diff
@@ -0,0 +1,222 @@
+        """
+        api:
+            String representing the type of request to be sent. In case the api
+            is missing or unrecognized the created object will be an ordinary
+            dictionary.
+        TODO:
+            To make the constructor fully dict like. E.g. I want to be able to
+            use it like:
+            ddmReq5 = DDMReqTemplate('copy',
+                                     item=['/LQLQToTop..._M-1200/NANOAODSIM'],
+                                     site=['T2_CH_CERN', 'T2_US_MIT'])
+        """
+        super(DDMReqTemplate, self).__init__(*arg, **kw)
+        if not api:
+            return
+        self.api = api
+
+        if api == 'copy':
+            self.update({'item': [],
+                         'site': ['T2_*', 'T1_*_Disk'],
+                         'group': 'DataOps',
+                         'n': None,
+                         'cache': None})
+
+        elif api == 'pollcopy':
+            self.update({'request_id': None,
+                         'item': None,
+                         'site': None,
+                         'status': None,
+                         'user': None})
+
+        elif api == 'cancelcopy':
+            self.update({'request_id': None})
+
+        else:
+            return
+
+    def strip(self):
+        """
+        _strip_
+
+        Method used to drop all the fields with value None. DDM does not
+        handle those quite well, and they better be removed from the request.
+        """
+
+        for key in list(self):
+            if not self[key]:
+                self.pop(key)
+
+    def isEqual(self, ddmReq, exclude=None):
+        """
+        _isEqual_
+
+        Equivalence method.
+        :exclude:
+            a key to be excluded from the comparison.
+        :return:
+            Bool
+        """
+        # first we check the apis - we cannot compare apples to oranges:
+        if not self.api == ddmReq.api:
+            return False
+
+        # in case no key have been excluded it is a normal dict comparison
+        if not exclude:
+            return self == ddmReq
+
+        # The final check: we iterate over the two objects separately in order
+        # to be sure we do not leave any extra key which is present in one of
+        # the objects but not in the other
+        for key in self:
+            if key in exclude:
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+
+        for key in ddmReq:
+            if key in exclude:
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+        return True
+
+class DDM(Service):
+    """
+    Class which provides client APIs to the DDM service.
+
+    Usage example:
+        api = 'copy'
+        ddmReq = DDMReqTemplate(api)
```

Comment:
it was the first implementation, but I changed it because it constrains you a lot. The idea is that you may create and manage the objects outside of the wrapper and deal with them through the logic of the service you are building while this one to stay a generic wrapper. Also you can keep track of the status outside of this code, through Mongodb or so. My original idea was(and still is) to build the list of workflows outside catch store the parameters for output in MongoDB marking them with to_DDM = true status or a separate field with more than a bool value like DDM_status = (pending | in progress| done)  and then with a simple cycle outside this code to go and create a list of ddmRequests with their relevant output paramenters and simply pass this list to makeAggrrequest and let it do the rest.

Reply:
And forgot to mention. This way we can also get rid of the usage of Phedex for the Relvals, and do everything through DDM instead. (we can put a site list in the request template and it will effectively be equivalent to a Phedex request, IIUC). It will make also the transition to Rucio much easier after that, because we will move the logic from a single place.

Reply:
Now all the parameters for the request may be passed directly to the constructor: 
```
ddmReq = DDMReqTemplate(api='pollcopy', request_id=46633)
```

Reply:
If functionality and latency is the same between a PhEDEx and DDM request, then I think we could indeed use DDM for all this output data placement. I'll write to Benedikt.

Reply:
@amaltaro I am not resolving this conversation here until we do not take a final decision. Please feel free to do so if you think it is better to continue the discussion offline.